src/components/Preferences/LocalIdentities.js

```diff
@@ -19,91 +20,149 @@ import {
 } from '../IdentityManager/IdentityManager'
 import EmptyLocalIdentities from './EmptyLocalIdentities'
 import Import from './Import'
+import { GU } from '../../utils'
 
 const iOS = /iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream
 
-const LocalIdentities = ({
-  dao,
-  localIdentities,
-  onClearAll,
-  onImport,
-  onModify,
-  onModifyEvent,
-}) => {
-  // transform localIdentities from object into array and attach address to each entry
-  const identities = Object.entries(localIdentities).map(
-    ([address, identity]) => ({
-      ...identity,
-      address,
+const LocalIdentities = React.memo(
+  ({ dao, localIdentities, onClearAll, onImport, onModify, onModifyEvent }) => {
+    if (!Object.keys(localIdentities).length) {
+      return <EmptyLocalIdentities onImport={onImport} />
+    }
+
+    const { identityEvents$ } = React.useContext(IdentityContext)
+    const { showLocalIdentityModal } = React.useContext(
+      LocalIdentityModalContext
+    )
+    const [addressesSelected, setAddressesSelected] = React.useState(
+      new WeakMap()
+    )
+    React.useEffect(() => {
+      const newAddressesSelected = new WeakMap()
+      Object.keys(localIdentities).forEach(key => {
+        newAddressesSelected.set(localIdentities[key], true)
+      })
+      setAddressesSelected(newAddressesSelected)
+    }, [localIdentities])
+    const updateLabel = React.useCallback(address => async () => {
+      try {
+        await showLocalIdentityModal(address)
+        // preferences get all
+        onModifyEvent()
+        // for iframe apps
+        identityEvents$.next({ type: identityEventTypes.MODIFY, address })
+      } catch (e) {
+        /* nothing was updated */
+      }
     })
-  )
+    const allSelected = Object.keys(localIdentities).every(key =>
+      addressesSelected.get(localIdentities[key])
+    )
+    const someSelected = Object.keys(localIdentities).some(key =>
+      addressesSelected.get(localIdentities[key])
+    )
 
-  if (!identities.length) {
-    return <EmptyLocalIdentities onImport={onImport} />
-  }
+    const handleToggleAllSelectedChange = React.useCallback(() => {
+      const newAddressesSelected = new WeakMap()
+      Object.keys(localIdentities).forEach(key =>
+        newAddressesSelected.set(
+          localIdentities[key],
+          !(allSelected || someSelected)
+        )
+      )
+      setAddressesSelected(newAddressesSelected)
+    })
+    const handleToggleAddressesSelected = React.useCallback(address => () => {
+      const newAddressesSelected = new WeakMap()
+      Object.keys(localIdentities).forEach(key => {
+        newAddressesSelected.set(
+          localIdentities[key],
+          address === key
+            ? !addressesSelected.get(localIdentities[key])
+            : addressesSelected.get(localIdentities[key])
+        )
+      })
+      setAddressesSelected(newAddressesSelected)
+    })
```

Comment:
For these two `useCallback()`s, should we add `localIdentities` and `setAddressesSelected` to their cached list?

Reply:
The `function` or the `WeakMap`?